File: src/main/java/com/examportal/controller/ExamController.java
package com.examportal.controller;

import java.security.Principal;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.examportal.dto.ApiResponse;
import com.examportal.dto.ExamDTO;
import com.examportal.dto.ExamResponseDTO;
import com.examportal.dto.ResultResponseDTO;
import com.examportal.dto.SubmitExamDTO;
import com.examportal.service.ExamService;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;

@CrossOrigin(origins = "*")
@RestController
@RequestMapping("/api/exams")

@Tag(name = "Exam Management", description = "Exam management APIs")
@SecurityRequirement(name = "bearerAuth")
public class ExamController {

    @Autowired
    private ExamService examService;

    @PostMapping
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Create a new exam (Teacher only)")
    public ResponseEntity<ExamResponseDTO> createExam(@Valid @RequestBody ExamDTO examDTO) {
        ExamResponseDTO exam = examService.createExam(examDTO);
        return ResponseEntity.status(HttpStatus.CREATED).body(exam);
    }

    @PutMapping("/{examId}")
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Update an exam (Teacher only)")
    public ResponseEntity<ExamResponseDTO> updateExam(@PathVariable Long examId, @Valid @RequestBody ExamDTO examDTO) {
        ExamResponseDTO exam = examService.updateExam(examId, examDTO);
        return ResponseEntity.ok(exam);
    }

    @DeleteMapping("/{examId}")
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Delete an exam (Teacher only)")
    public ResponseEntity<ApiResponse> deleteExam(@PathVariable Long examId) {
        examService.deleteExam(examId);
        return ResponseEntity.ok(new ApiResponse("Exam deleted successfully"));
    }

    @GetMapping("/{examId}")
    @PreAuthorize("hasRole('STUDENT') or hasRole('TEACHER')")
    @Operation(summary = "Get exam by ID")
    public ResponseEntity<ExamResponseDTO> getExamById(@PathVariable Long examId) {
        ExamResponseDTO exam = examService.getExamById(examId);
        return ResponseEntity.ok(exam);
    }

    @GetMapping
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Get all exams (Teacher only)")
    public ResponseEntity<List<ExamResponseDTO>> getAllExams() {
        List<ExamResponseDTO> exams = examService.getAllExams();
        return ResponseEntity.ok(exams);
    }

    @GetMapping("/active")
    @Operation(summary = "Get active exams")
    public ResponseEntity<List<ExamResponseDTO>> getActiveExams() {
        List<ExamResponseDTO> exams = examService.getActiveExams();
        return ResponseEntity.ok(exams);
    }

    @GetMapping("/upcoming")
    @Operation(summary = "Get upcoming exams")
    public ResponseEntity<List<ExamResponseDTO>> getUpcomingExams() {
        List<ExamResponseDTO> exams = examService.getUpcomingExams();
        return ResponseEntity.ok(exams);
    }

    @GetMapping("/subject")
    @Operation(summary = "Get exams by subject")
    public ResponseEntity<List<ExamResponseDTO>> getExamsBySubject(@RequestParam String subject) {
        List<ExamResponseDTO> exams = examService.getExamsBySubject(subject);
        return ResponseEntity.ok(exams);
    }

    @PostMapping("/submit")
    @PreAuthorize("hasRole('STUDENT')")
    @Operation(summary = "Submit exam answers (Student only)")
    public ResponseEntity<ResultResponseDTO> submitExam(Principal principal, @Valid @RequestBody SubmitExamDTO submitExamDTO) {
        ResultResponseDTO result = examService.submitExam(principal.getName(), submitExamDTO);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/results/my")
    @PreAuthorize("hasRole('STUDENT')")
    @Operation(summary = "Get my exam results (Student only)")
    public ResponseEntity<List<ResultResponseDTO>> getMyResults(Principal principal) {
        List<ResultResponseDTO> results = examService.getResultsByUser(principal.getName());
        return ResponseEntity.ok(results);
    }

    @GetMapping("/{examId}/results")
    @PreAuthorize("hasRole('TEACHER')")
    @Operation(summary = "Get results for an exam (Teacher only)")
    public ResponseEntity<List<ResultResponseDTO>> getResultsByExam(@PathVariable Long examId) {
        List<ResultResponseDTO> results = examService.getResultsByExam(examId);
        return ResponseEntity.ok(results);
    }
}

File: Dotnet3/ExamPortal/Controllers/ExamController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using ExamPortal.DTOs;
using ExamPortal.Services.Interfaces;
using System.Security.Claims;

namespace ExamPortal.Controllers
{
    [ApiController]
    [Route("api/exams")]
    [Authorize]
    public class ExamController : ControllerBase
    {
        private readonly IExamService _examService;
        
        public ExamController(IExamService examService)
        {
            _examService = examService;
        }
        
        /// <summary>
        /// Create a new exam (Teacher only)
        /// </summary>
        [HttpPost]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<ExamResponseDTO>> CreateExam([FromBody] ExamDTO examDTO)
        {
            var exam = await _examService.CreateExamAsync(examDTO);
            return CreatedAtAction(nameof(GetExamById), new { examId = exam.Id }, exam);
        }
        
        /// <summary>
        /// Update an exam (Teacher only)
        /// </summary>
        [HttpPut("{examId}")]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<ExamResponseDTO>> UpdateExam(long examId, [FromBody] ExamDTO examDTO)
        {
            var exam = await _examService.UpdateExamAsync(examId, examDTO);
            return Ok(exam);
        }
        
        /// <summary>
        /// Delete an exam (Teacher only)
        /// </summary>
        [HttpDelete("{examId}")]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<ApiResponse>> DeleteExam(long examId)
        {
            await _examService.DeleteExamAsync(examId);
            return Ok(new ApiResponse("Exam deleted successfully"));
        }
        
        /// <summary>
        /// Get exam by ID
        /// </summary>
        [HttpGet("{examId}")]
        [Authorize(Roles = "STUDENT,TEACHER")]
        public async Task<ActionResult<ExamResponseDTO>> GetExamById(long examId)
        {
            var exam = await _examService.GetExamByIdAsync(examId);
            return Ok(exam);
        }
        
        /// <summary>
        /// Get all exams (Teacher only)
        /// </summary>
        [HttpGet]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<List<ExamResponseDTO>>> GetAllExams()
        {
            var exams = await _examService.GetAllExamsAsync();
            return Ok(exams);
        }
        
        /// <summary>
        /// Get active exams
        /// </summary>
        [HttpGet("active")]
        public async Task<ActionResult<List<ExamResponseDTO>>> GetActiveExams()
        {
            var exams = await _examService.GetActiveExamsAsync();
            return Ok(exams);
        }
        
        /// <summary>
        /// Get upcoming exams
        /// </summary>
        [HttpGet("upcoming")]
        public async Task<ActionResult<List<ExamResponseDTO>>> GetUpcomingExams()
        {
            var exams = await _examService.GetUpcomingExamsAsync();
            return Ok(exams);
        }
        
        /// <summary>
        /// Get exams by subject
        /// </summary>
        [HttpGet("subject")]
        public async Task<ActionResult<List<ExamResponseDTO>>> GetExamsBySubject([FromQuery] string subject)
        {
            var exams = await _examService.GetExamsBySubjectAsync(subject);
            return Ok(exams);
        }
        
        /// <summary>
        /// Submit exam answers (Student only)
        /// </summary>
        [HttpPost("submit")]
        [Authorize(Roles = "STUDENT")]
        public async Task<ActionResult<ResultResponseDTO>> SubmitExam([FromBody] SubmitExamDTO submitExamDTO)
        {
            var username = User.FindFirst(ClaimTypes.Name)?.Value;
            if (string.IsNullOrEmpty(username))
                return Unauthorized();
                
            var result = await _examService.SubmitExamAsync(username, submitExamDTO);
            return Ok(result);
        }
        
        /// <summary>
        /// Get my exam results (Student only)
        /// </summary>
        [HttpGet("results/my")]
        [Authorize(Roles = "STUDENT")]
        public async Task<ActionResult<List<ResultResponseDTO>>> GetMyResults()
        {
            var username = User.FindFirst(ClaimTypes.Name)?.Value;
            if (string.IsNullOrEmpty(username))
                return Unauthorized();
                
            var results = await _examService.GetResultsByUserAsync(username);
            return Ok(results);
        }
        
        /// <summary>
        /// Get results for an exam (Teacher only)
        /// </summary>
        [HttpGet("{examId}/results")]
        [Authorize(Roles = "TEACHER")]
        public async Task<ActionResult<List<ResultResponseDTO>>> GetResultsByExam(long examId)
        {
            var results = await _examService.GetResultsByExamAsync(examId);
            return Ok(results);
        }
    }
}
